org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/MethodBinding.java

```diff
@@ -603,6 +605,32 @@ public IVariableBinding[] getSyntheticOuterLocals() {
 		public  void setSyntheticOuterLocals(IVariableBinding[] syntheticOuterLocalVariables) {
 			this.syntheticOuterLocalVariables = syntheticOuterLocalVariables;
 		}
+
+		@Override
+		protected JavaElement getUnresolvedJavaElement() {
+			if (JavaCore.getPlugin() == null) {
+				return null;
+			}
+			if (!(this.resolver instanceof DefaultBindingResolver)) return null;
+	
+			DefaultBindingResolver defaultBindingResolver = (DefaultBindingResolver) this.resolver;
+			if (!defaultBindingResolver.fromJavaProject) return null;
+			ASTNode domNode = (ASTNode)defaultBindingResolver.bindingsToAstNodes.get(this);
+			// resolving all parent lambdas is necessary for proper resolution
+			// as it populates newAstToOldAst.
+			ASTNode current = domNode;
+			while (current != null) {
```

Comment:
Does this need to be done every time or can it be done once and then marked as completed or can the LambdaFactory LambdaMethod be cached to avoid this in the future?

Reply:
Usage of this method is relatively sporadic (happens only when user Ctrl+clicks it at the moment) and execution remains relatively fast. On the other hand, the DOM may be retained for a long time, so adding some cache here increase memory consumption.
Overall, I think it's best to not cache and keep more memory free for that case. We can easily revisit later if my impressions are proven wrong though.